api/v1/appsource_types.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// AppSourceSpec defines the desired state of AppSource
+type AppSourceSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of AppSource. Edit appsource_types.go to remove/update
+	Path    string `json:"path,string"`
+	RepoURL string `json:"repoURL,string"`
+}
+
+// AppSourceStatus defines the observed state of AppSource
+type AppSourceStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// AppSource is the Schema for the appsources API
+type AppSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   AppSourceSpec   `json:"spec,omitempty"`
+	Status AppSourceStatus `json:"status,omitempty"`
+}
+
+func (a *AppSource) ApplicationFromSource(req ctrl.Request) *v1alpha1.Application {
+	return &v1alpha1.Application{
+		ObjectMeta: metav1.ObjectMeta{Name: req.Name},
+		Spec: v1alpha1.ApplicationSpec{
+			Source: v1alpha1.ApplicationSource{
+				RepoURL: a.Spec.RepoURL,
+				Path:    a.Spec.Path,
+			},
+			Project: req.Namespace,
```

Comment:
AppSource namespace should not be used as a project name. The project name is a part of namespace name as configured in `project.pattern` setting.  So controller should run  `project.pattern` regex and extract project name from first capturing group.

Reply:
Will it always be the first capturing group or should we use a named capturing group? I think we could also configure it to look for the named project capture group, and if nothing is found it defaults to the first named capture group